### PR TITLE
Improvements to `$ocaml_cv_cc_vendor` in configure

### PR DESCRIPTION
**Pull request by MisterDA**
This PR improves the detection of the C compiler on Windows, and replaces uses of `$cc_basename` by our `$ocaml_cv_cc_vendor`. More specifically, it:

1. Allows to properly discover whether mingw-w64 is used (`$ocaml_cv_cc_vendor` used to conflate it with GCC), and whether mingw-w64 is coupled to GCC or Clang. The build with mingw-w64+clang is currently rejected (but a follow-up PR with a couple of patches might appear on the net). I've chosen the `mingw32` prefix, we can go with another. I haven't found a case where a previous match for "gcc-\*" that wouldn't match after this PR (because the match is now "mingw32-\*") would be problematic, apart from removing `-latomic` on i686 that apparently has no effect.
2. Detects clang-cl, a clang that's [ABI compatible](https://clang.llvm.org/docs/MSVCCompatibility.html) and [command-line compatible](https://clang.llvm.org/docs/UsersManual.html#clang-cl) with MSVC. Considering that we're interested in it first and foremost for its compatibility with MSVC, I've chosen to make it report as "msvc-xx-clang-yy" rather than "clang-yy-msvc-xx" so that configure scripts for MSVC also apply to this compiler. 
3. Fixes TSan detection on macOS: configure would look for "clang" in the cc basename and would fail, because the default compiler is "cc". This would also happen if one were to run `./configure --enable-tsan CC=cc`, even if `cc` supported TSan.
4. Removes uses of `$cc_basename` as it is an internal and undocumented feature of Autoconf, and prefer our `$ocaml_cv_cc_vendor` instead. This is prompted by point 3, and I hope it'll prevent future problems.

Accurate reports of the C compiler vendor also improve the quality of bug reports! "What C compiler are you using on Windows?" is now "oh configure has picked this compiler at this version".

### CODE REVIEW

File: aclocal.m4
Comment (ghost) [inline aclocal.m4:49]: Looking at the mingw block below, it seems it is acceptable for this macro to return its output on multiple lines, so why not use the same logic for msvc? I.e.
```
#if defined(_MSC_VER)
msvc _MSC_VER
#if defined(clang stuff)
clang suffix
#endif
#endif
```

Comment (MisterDA) [reply aclocal.m4:49]: True! Thanks for the suggestion.

**Comment by MisterDA** [comment]
Maybe "mingw32" is confusing, and "mingw" would be more appropriate?

**Comment by shindere** [comment]
Antonin Décimo (2023/11/23 01:51 -0800):
> Maybe "mingw32" is confusing, and "mingw" would be more appropriate?

Yes! And congrats, that looks like great work to me! Thanks!

**Comment by MisterDA** [comment]
>> Maybe "mingw32" is confusing, and "mingw" would be more appropriate?
>
> Yes! And congrats, that looks like great work to me! Thanks!

Done, and thanks :)

**Comment by MisterDA** [comment]
I've rebased this PR and added a Changes entry.

**Comment by OlivierNicole** [comment]
Is something holding the PR from being merged? Speaking egoistically, it would be great to have it in 5.2 together with TSan as it fixes the configure step on some macOS machines with `--enable-tsan`. (So the “Windows” label is a bit misleading.)

**Comment by ghost** [comment]
> Is something holding the PR from being merged? Speaking egoistically, it would be great to have it in 5.2 together with TSan as it fixes the configure step on some macOS machines with `--enable-tsan`. (So the “Windows” label is a bit misleading.)

I think the part which matters for TSan on macOS got stealthly merged as part of the TSan-on-FreeBSD support. This doesn't make this PR irrelevant, of course.

**Comment by MisterDA** [comment]
> I think the part which matters for TSan on macOS got stealthly merged as part of the TSan-on-FreeBSD support. 

0c82f6354111fd235f02583686e15829074b5b19

Reading old configure code, I think we have the same detection issue, but for frame pointers… fixed by this PR.

**Review by ghost — APPROVED** [review]

**Comment by MisterDA** [comment]
Rebased and reworded the commit messages to account for the fixes that went in trunk in the meantime.

**Comment by gasche** [comment]
> Is something holding the PR from being merged? 

At the time you asked, the PR was waiting for an approval. (There are not that many people who review `configure` change, typically it is @shindere or @dra27.) Since then @dustanddreams approved, so I can merge.

**Review by gasche — APPROVED** [review]
Approved on behalf of @dustanddreams.

**Comment by OlivierNicole** [comment]
I believe @shindere had approved it also.

**Comment by gasche** [comment]
Not that I can see. (Also @shindere has merge rights so, if he had approved, he would probably either have merged or indicated what remains to be done before merging.)

Note: approving is more than saying "this works looks nice", it means that the person is confident that we will not regret merging the PR, which generally implies/requires a detailed review.

**Comment by OlivierNicole** [comment]
OK, my bad, I probably misinterpreted.

**Comment by shindere** [comment]
I think I messed up here by not being explicit on whether I would do a
full review or not and I would like to apologize for that.

I didn't feel completely confident in my ability to review the proposed
code but I should have been explicit on that, precisely.

Will do better next time. I believe that the PR does improve the
situation over what was there before.

**Pull request merged**
